File: modules/pygauss/src/random.cpp
/* Copyright (c) 2021 Grumpy Cat Software S.L.
 *
 * This Source Code is licensed under the MIT 2.0 license.
 * the terms can be found in LICENSE.md at the root of
 * this project, or at http://mozilla.org/MPL/2.0/.
 */

#include <arrayfire.h>
#include <pybind11/pybind11.h>
#include <pybind11/stl.h>
#include <gauss.h>
#include <pygauss.h>

#define FLOAT32_EPS 1.1920929e-07

namespace py = pybind11;

void pygauss::bindings::random_numbers(py::module &m) {

    py::enum_<af::randomEngineType>(m, "RandomEngineType", "Built-in engines for random number generation")
            .value("Default", af::randomEngineType::AF_RANDOM_ENGINE_DEFAULT, "Defaults to Philox")
            .value("Mersenne", af::randomEngineType::AF_RANDOM_ENGINE_MERSENNE, "Mersenne GP 11213.")
            .value("Threefry", af::randomEngineType::AF_RANDOM_ENGINE_THREEFRY, "Threefry 2X32_16.")
            .value("Philox", af::randomEngineType::AF_RANDOM_ENGINE_PHILOX, "Philox 4x32_10.")
            .export_values();

    py::class_<af::randomEngine>(m, "ShapeletsRandomEngine")
            .def("uniform",
                 [](af::randomEngine &self, const double low, const double high, const af::dim4 &shape,
                    const af::dtype &dtype) {
                     return gauss::random::uniform(low, high, shape, dtype, self);
                 },
                 py::arg("low") = 0.0,
                 py::arg("high") = 1.0,
                 py::arg("shape") = af::dim4(1, 1, 1, 1),
                 py::arg("dtype") = af::dtype::f32,
                 "Samples are uniformly distributed over the half-open interval [low, high)")

            .def("exponential",
                 [](af::randomEngine &self, const double scale, const af::dim4 &shape, const af::dtype &dtype) {
                     return gauss::random::exponential(scale, shape, dtype, self);
                 },
                 py::arg("scale") = 1.0,
                 py::arg("shape") = af::dim4(1, 1, 1, 1),
                 py::arg("dtype") = af::dtype::f32,
                 "Draw samples from an exponential distribution.  Scale is the inverse of lambda.")

            .def("gamma",
                 [](af::randomEngine &self, const double alpha, const double scale,
                    const af::dim4 &shape, const af::dtype &dtype) {
                     return gauss::random::gamma(alpha, 1.0 / scale, shape, dtype, self);
                 },
                 py::arg("alpha").none(false),
                 py::arg("scale") = 1.0,
                 py::arg("shape") = af::dim4(1, 1, 1, 1),
                 py::arg("dtype") = af::dtype::f32,
                 "Draw samples from a Gamma distribution.  Alpha is what is called "
                 "shape or K parameter of the Gamma distribution")

            .def("chisquare",
                 [](af::randomEngine &self, double df, const af::dim4 &shape, const af::dtype &dtype) {
                     return gauss::random::chisquare(df, shape, dtype, self);
                 },
                 py::arg("df").none(false),
                 py::arg("shape") = af::dim4(1, 1, 1, 1),
                 py::arg("dtype") = af::dtype::f32,
                 "Draw samples from a chi-square distribution.  df is the degree of freedom")

            .def("beta",
                 [](af::randomEngine &self, const double a, const double b, const af::dim4 &shape,
                    const af::dtype &dtype) {
                     return gauss::random::beta(a, b, shape, dtype, self);
                 },
                 py::arg("a").none(false),
                 py::arg("b").none(false),
                 py::arg("shape") = af::dim4(1, 1, 1, 1),
                 py::arg("dtype") = af::dtype::f32,
                 "Draw samples from a Beta distribution.  Alpha is what is called "
                 "shape or K parameter of the Gamma distribution")

            .def("wald",
                 [](af::randomEngine &self, const double mean, const double scale, const af::dim4 &shape,
                    const af::dtype &dtype) {
                     return gauss::random::wald(mean, scale, shape, dtype, self);
                 },
                 py::arg("mean").none(false),
                 py::arg("scale").none(false),
                 py::arg("shape") = af::dim4(1, 1, 1, 1),
                 py::arg("dtype") = af::dtype::f32,
                 "Draw samples from a Wald, or inverse Gaussian, distribution.")

            .def("normal",
                 [](af::randomEngine &self, const double mean, const double sigma, const af::dim4 &shape,
                    const af::dtype &dtype) {
                     return gauss::random::normal(mean, sigma, shape, dtype, self);
                 },
                 py::arg("mean") = 0.0,
                 py::arg("sigma") = 1.0,
                 py::arg("shape") = af::dim4(1, 1, 1, 1),
                 py::arg("dtype") = af::dtype::f32,
                 "Draw random samples from a normal (Gaussian) distribution.")

            .def("standard_normal",
                 [](af::randomEngine &self, const af::dim4 &shape, const af::dtype &dtype) {
                     return gauss::random::normal(0.0, 1.0, shape, dtype, self);
                 },
                 py::arg("shape") = af::dim4(1, 1, 1, 1),
                 py::arg("dtype") = af::dtype::f32,
                 "Draw random samples from a normal (Gaussian) distribution.")

            .def("lognormal",
                 [](af::randomEngine &self, const double mean, const double sigma, const af::dim4 &shape,
                    const af::dtype &dtype) {
                     return gauss::random::lognormal(mean, sigma, shape, dtype, self);
                 },
                 py::arg("mean") = 0.0,
                 py::arg("sigma") = 1.0,
                 py::arg("shape") = af::dim4(1, 1, 1, 1),
                 py::arg("dtype") = af::dtype::f32,
                 "Draw samples from a log-normal distribution.")

            .def("logistic",
                 [](af::randomEngine &self, const double loc, const double scale, const af::dim4 &shape,
                    const af::dtype &dtype) {
                     return gauss::random::logistic(loc, scale, shape, dtype, self);
                 },
                 py::arg("loc") = 0.0,
                 py::arg("scale") = 1.0,
                 py::arg("shape") = af::dim4(1, 1, 1, 1),
                 py::arg("dtype") = af::dtype::f32,
                 "Draw samples from a logistic distribution.")

            .def("multivariate_normal",
                 [](af::randomEngine &self, const py::object &mean, const py::object &cov,
                    const int64_t samples, const af::dtype &dtype) {

                     auto m = arraylike::as_array_checked(mean).as(dtype);
                     auto cv = arraylike::as_array_checked(cov).as(dtype);

                     return gauss::random::multivariate_normal(samples, m, cv, self);
                 },
                 py::arg("mean").none(false),
                 py::arg("cov").none(false),
                 py::arg("samples").none(false),
                 py::arg("dtype") = af::dtype::f32,
                 "Draw samples from a logistic distribution.")
            .def("permutation",
                 [](af::randomEngine &self, const py::object &x, const int32_t axis) -> af::array {
                     auto is_scalar = arraylike::is_scalar(x);
                     auto original = is_scalar ?
                                     af::iota(af::dim4(py::cast<dim_t>(x)), af::dim4(1), af::dtype::s32) :
                                     arraylike::as_array_checked(x);

                     auto checked_axis = is_scalar ? 0 : axis;
                     return gauss::random::permute(original, checked_axis, self);
                 },
                 py::arg("x").none(false),
                 py::arg("axis") = 0);

    m.def("default_rng",
          [](const af::randomEngineType type, const unsigned long long seed) {
              return af::randomEngine(type, seed);
          },
          py::arg("type") = af::randomEngineType::AF_RANDOM_ENGINE_DEFAULT,
          py::arg("seed") = 0ULL);

    m.def("permutation",
          [](const py::object &x, const int32_t axis, std::optional<af::randomEngine> &engine) -> af::array {
              auto is_scalar = arraylike::is_scalar(x);
              auto original = is_scalar ?
                              af::iota(af::dim4(py::cast<dim_t>(x)), af::dim4(1), af::dtype::s32) :
                              arraylike::as_array_checked(x);

              auto checked_axis = is_scalar ? 0 : axis;
              return gauss::random::permute(original, checked_axis, engine);
          },
          py::arg("x").none(false),
          py::arg("axis") = 0,
          py::arg("engine") = py::none());

    m.def("random",
          [](const af::dim4 &shape, const af::dtype &dtype, std::optional<af::randomEngine> &engine) {
              return gauss::random::uniform(0.0, 1.0, shape, dtype, engine);
          },
          py::arg("shape").none(false),
          py::arg("dtype") = af::dtype::f32,
          py::arg("engine") = py::none());

    m.def("randn", [](const af::dim4 &shape, const af::dtype &dtype, std::optional<af::randomEngine> &engine) {
              return gauss::random::normal(0.0, 1.0, shape, dtype, engine);
          },
          py::arg("shape").none(false),
          py::arg("dtype") = af::dtype::f32,
          py::arg("engine") = py::none());

    m.def("randint",
          [](const int64_t low, const std::optional<int64_t> high, const af::dim4 &shape, const af::dtype &dtype,
             std::optional<af::randomEngine> &engine) {
              return gauss::random::randint(low, high, shape, dtype, engine);
          },
          py::arg("low").none(false),
          py::arg("high") = py::none(),
          py::arg("shape") = af::dim4(1, 1),
          py::arg("dtype") = af::dtype::s32,
          py::arg("engine") = py::none());
}

File: modules/shapelets/compute/random.py
from __future__ import annotations
from typing import Optional
try:
    from typing import Literal
except ImportError:
    from typing_extensions import Literal

from .__basic_typing import ShapeLike, DataTypeLike, ArrayLike
from ._array_obj import ShapeletsArray

from ._pygauss import (
    default_rng as __default_rng,
    randint as __randint,
    randn as __randn,
    random as __random,
    permutation as __permutation,
    ShapeletsRandomEngine,
    RandomEngineType as __RandomEngineType
)

class ShapeletsRandomEngine:
    """
    Wrapper class around a random number algorithm suitable to be used in GPU / OpenCL computations

    Use :obj:`~shapelets.compute.random.random_engine` as a constructor function, where it is possible 
    to specify the underlyting random generator algorithm as well as the a seed value for controllable 
    results.
    """

    def beta(self, a: float, b: float, shape: ShapeLike = (1, 1), dtype: DataTypeLike = 'float32') -> ShapeletsArray:
        r"""
        Return samples from the beta distribution.

        The calculation is based on the Gamma distribution, in such way that:

        .. math::

            \\\begin{matrix} X \sim \Gamma(\alpha, \theta) \\ Y \sim \Gamma(\beta, \theta) \end{matrix} \longrightarrow \frac{X}{X+Y} \sim B(\alpha,\beta)

        Parameters
        ----------
        a: float
            It is normally called shape or also known as the `K` parameter of the gamma distribution.        
        b: float, defaults to 1.0
            The scale parameter.
        shape: ShapeLike, defaults to (1,1)
            Dimensions of the tensor.
        dtype: DataTypeLike, defaults to 'float32'
            Type of the resulting elements

        Returns
        -------
        ShapeletsArray
            A new array instance whose values are drawn from the beta distribution.   

        """

    def chisquare(self, df: float, shape: ShapeLike = (1, 1), dtype: DataTypeLike = 'float32') -> ShapeletsArray:
        r"""
        Return samples from the chi square distribution.

        The calculation is based on the Gamma distribution, in such way that:

        .. math::

            X \sim \Gamma(\frac{df}{2},2) \longrightarrow \chi^2(df)

        Parameters
        ----------
        df: float
            Degrees of freedom.        
        shape: ShapeLike, defaults to (1,1)
            Dimensions of the tensor.
        dtype: DataTypeLike, defaults to 'float32'
            Type of the resulting elements

        Returns
        -------
        ShapeletsArray
            A new array instance whose values are drawn from the chi square distribution.  
        """

    def exponential(self, scale: float = 1.0, shape: ShapeLike = (1, 1), dtype: DataTypeLike = 'float32') -> ShapeletsArray:
        r"""
        Return samples from the exponential distribution.

        The calculation is based on the Uniform U(0,1) distribution, in such way that:

        .. math::

            X \sim U(0,1) \longrightarrow \frac{-\log(1 - X)}{\lambda} \sim E(\lambda)

        Parameters
        ----------
        scale: float
            Inverse of lambda.        
        shape: ShapeLike, defaults to (1,1)
            Dimensions of the tensor.
        dtype: DataTypeLike, defaults to 'float32'
            Type of the resulting elements

        Returns
        -------
        ShapeletsArray
            A new array instance whose values are drawn from the exponential distribution.  
        """

    def gamma(self, alpha: float, scale: float = 1.0, shape: ShapeLike = (1, 1), dtype: DataTypeLike = 'float32') -> ShapeletsArray:
        r"""
        Return samples from the gamma distribution.

        The implementation is based on the algorithm 4.33 found in reference [1]

        Parameters
        ----------
        alpha: float
            It is normally called shape or also known as the `K` parameter of the gamma distribution.        
        scale: float, defaults to 1.0
            The scale parameter.
        shape: ShapeLike, defaults to (1,1)
            Dimensions of the tensor.
        dtype: DataTypeLike, defaults to 'float32'
            Type of the resulting elements

        Returns
        -------
        ShapeletsArray
            A new array instance whose values are drawn from the gamma distribution.            

        References
        ----------
        [1] Base line implementation as defined in: 
            `Handbook of Monte Carlo Methods <http://www.maths.uq.edu.au/~kroese/montecarlohandbook>`_
            D.P. Kroese, T. Taimre, Z.I. Botev. John Wiley & Sons, 2011.

        """

    def logistic(self, loc: float = 0.0, scale: float = 1.0, shape: ShapeLike = (1, 1), dtype: DataTypeLike = 'float32') -> ShapeletsArray:
        r"""
        Return samples from the logistic distribution.

        The implementation is based on uniform distribution, transforming it the following way:

        .. math::

            X \sim U(0, 1) \longrightarrow \mu + \beta(\log(X) - \log(1 - X)) \sim Logistic(\mu,\beta)

        where ``loc`` represents mu and ``scale`` is represented by beta.

        Parameters
        ----------
        loc: float, defaults to 0.0
            Location parameter
        scale: float, defaults to 1.0
            Scale parameter
        shape: ShapeLike, defaults to (1,1)
            Dimensions of the tensor.
        dtype: DataTypeLike, defaults to 'float32'
            Type of the resulting elements

        Returns
        -------
        ShapeletsArray
            A new array instance whose values are drawn from the logistic distribution.
        """

    def lognormal(self, mean: float = 0.0, sigma: float = 1.0, shape: ShapeLike = (1, 1), dtype: DataTypeLike = 'float32') -> ShapeletsArray:
        r"""
        Return samples from the lognormal distribution.

        Based on the standard normal distribution, the lognormal is calculated through:

        .. math::

            Z \sim N(0,1) \longrightarrow e^{\mu + \sigma Z} \sim LogNormal(\mu, \sigma)

        Parameters
        ----------
        mean: float, defaults to 0.0
            Expected value (:math:`\mu`)
        sigma: float, defaults to 1.0
            Standard deviation of the variable's natural logaritm (:math:`\sigma`)
        shape: ShapeLike, defaults to (1,1)
            Dimensions of the tensor.
        dtype: DataTypeLike, defaults to 'float32'
            Type of the resulting elements

        Returns
        -------
        ShapeletsArray
            A new array instance whose values are drawn from the lognormal distribution.
        """

    def multivariate_normal(self, mean: ArrayLike, cov: ArrayLike, samples: int, dtype: DataTypeLike = 'float32') -> ShapeletsArray:
        r"""
        Draw random samples from a multivariate normal distribution.

        A multivariate normal distribution is specified by its mean and covariance matrix. 

        Parameters
        ----------
        mean: 1-D array of length N
            Mean of each one of the series to generate

        cov: 2-D array of shape NxN
            Symmetric and positive-semidefinite covariance matrix
        
        samples: int
            Determines the lenght of each series to generate.  
        
        dtype: DataTypeLike, defaults to 'float32'
            Type of the resulting elements            

        Returns
        -------
        ShapeletsArray
            A new array, of dimensions Nxsamples. 

        """

    def normal(self, mean: float = 0.0, sigma: float = 1.0, shape: ShapeLike = (1, 1), dtype: DataTypeLike = 'float32') -> ShapeletsArray:
        r"""
        Draw random samples from a normal (Gaussian) distribution.

        Parameters
        ----------
        mean: float, defaults to 0.0
            Mean of the distribution

        sigma: float, defaults to 1.0
            Standard deviation of the distribution
        
        shape: ShapeLike, defaults to (1,1)
            Dimensions of the tensor.
            
        dtype: DataTypeLike, defaults to 'float32'
            Type of the resulting elements          

        Returns
        -------
        ShapeletsArray
            A new array instance whose values are drawn from the gaussian distribution.

        """

    def standard_normal(self, shape: ShapeLike = (1, 1), dtype: DataTypeLike = 'float32') -> ShapeletsArray:
        r"""
        Draw random samples from a standard normal (Gaussian) distribution.

        A standard gaussian distribution is one with mean 0.0 and 1.0 standard deviation.

        Parameters
        ----------
        shape: ShapeLike, defaults to (1,1)
            Dimensions of the tensor.
            
        dtype: DataTypeLike, defaults to 'float32'
            Type of the resulting elements          

        Returns
        -------
        ShapeletsArray
            A new array instance whose values are drawn from the gaussian distribution.

        """

    def uniform(self, low: float = 0.0, high: float = 1.0, shape: ShapeLike = (1, 1), dtype: DataTypeLike = 'float32') -> ShapeletsArray:
        r"""
        Draw samples from a uniform distribution.

        Parameters
        ----------
        low: float, defaults to 0.0
            Lower bound for the output.  All values will be greater than or equal to low.

        high: float, defaults to 1.0
            Upper bound for the output.  All values will be less than high.

        shape: ShapeLike, defaults to (1,1)
            Dimensions of the tensor.
            
        dtype: DataTypeLike, defaults to 'float32'
            Type of the resulting elements          

        Returns
        -------
        ShapeletsArray
            A new array instance whose values are drawn from the uniform distribution.
        """

    def wald(self, mean: float, scale: float, shape: ShapeLike = (1, 1), dtype: DataTypeLike = 'float32') -> ShapeletsArray:
        r"""
        Return samples from the Wald distribution, also known as the inverse Gaussian distribution.

        Parameters
        ----------
        mean: float, defaults to 0.0
            Expected value 
        scale: float, defaults to 1.0
            Also known as lambda parameter. 
        shape: ShapeLike, defaults to (1,1)
            Dimensions of the tensor.
        dtype: DataTypeLike, defaults to 'float32'
            Type of the resulting elements

        Returns
        -------
        ShapeletsArray
            An new array instance whose values are drawn from the inverse Gaussian distribution.


        """

    def permutation(self, x: ArrayLike, axis=0) -> ShapeletsArray:
        r"""
        Randomly shuffles an array around a particular axis.

        Parameters
        ----------
        x: ArrayLike.  Defaults to (1,1).
            Input array

        axis: int.  Defaults to 0
            Axis to be shuffled

        Returns
        -------
        ShapeletsArray
            A randomly suffled array.
        """


RandomEngineType = Literal['mersenne', 'philox', 'threefry', 'default']


def random_engine(type: RandomEngineType = 'default', seed: int = 0) -> ShapeletsRandomEngine:
    """
    Constructs a new :obj:`~shapelets.compute.random.ShapeletsRandomEngine`.

    Parameters
    ----------
    type: RandomEngineType.  
        Selects the random generator algorithm. 

    seed: int.  Defaults to zero
        Seed value for initialization.

    Returns
    -------
    ShapeletsRandomEngine
        A new instance of a random engine with the selected algorithm and seed.

    Notes
    -------
    There are 3 different engines to choose from, where ``default`` value 
    maps to ``philox`` engine: 

    +----------+-------------------+---------+
    | Engine   | Implementation    | Default |
    +==========+===================+=========+
    | philox   | Philox 4x32-10    |   Yes   |
    +----------+-------------------+---------+
    | threefry | Threefry 2X32-16  |         |
    +----------+-------------------+---------+
    | mersenne | Mersenne GP 11213 |         |
    +----------+-------------------+---------+

    References
    ----------
    [1] Parallel random numbers: As easy as 1, 2, 3.  
        John K. Salmon, Mark A. Moraes, Ron O. Dror, and David E. Shaw.

        In Proceedings of 2011 International Conference for High Performance Computing, Networking, Storage and Analysis.
        DOI:`10.1145/2063384.2063405 <https://dl.acm.org/doi/10.1145/2063384.2063405>`_

    """
    native_ret = __RandomEngineType.Default
    if type == 'mersenne':
        native_ret = __RandomEngineType.Mersenne
    elif type == 'philox':
        native_ret = __RandomEngineType.Philox
    elif type == 'threefry':
        native_ret = __RandomEngineType.Threefry
    elif type != 'default':
        raise ValueError("Unknown random engine type")

    return __default_rng(native_ret, seed)


def randn(shape: ShapeLike, dtype: DataTypeLike = 'float32', engine: Optional[ShapeletsRandomEngine] = None) -> ShapeletsArray:
    """
    Creates an array whose elements are drawn from the standard normal distribution ``N(0,1)``.

    Parameters
    ----------
    shape: ShapeLike.  Defaults to (1,1).
        Shape of the array to be returned.

    dtype: DataTypeLike.  Defaults to float32.
        Final type of elements of the array.

    engine: ShapeletsRandomEngine.  Defaults to None.
        When not set, uses the default, implicit, random engine algorithm.  However,
        when it is set, this method is equivalent to call 
        :obj:`~shapelets.compute.random.ShapeletsRandomEngine.standard_normal` on the engine.

    Returns
    -------
    ShapeletsArray
        An new array instance whose values are drawn from the standard normal distribution.

    See Also
    --------
    :obj:`~shapelets.compute.random.ShapeletsRandomEngine`

    Examples
    --------
    Three by three array of samples from N(2,9):

    >>> import shapelets.compute as sc
    >>> 2.0 + 3 * sc.random.randn((3,3))
    [3 3 1 1]
    -0.7740     3.0547    -0.3060 
    2.5425     0.9645     2.7238 
    9.6323     2.6574    -1.5845 

    """
    return __randn(shape, dtype, engine)


def random(shape: ShapeLike, dtype: DataTypeLike = 'float32', engine: Optional[ShapeletsRandomEngine] = None) -> ShapeletsArray:
    """
    Creates an array whose elements are drawn from the continuous uniform distribution in the interval `[0.0, 1.0)` 

    Parameters
    ----------
    shape: ShapeLike.  Defaults to (1,1).
        Shape of the array to be returned.

    dtype: DataTypeLike.  Defaults to float32.
        Final type of elements of the array.

    engine: ShapeletsRandomEngine.  Defaults to None.
        When not set, uses the default, implicit, random engine algorithm.  However,
        when it is set, this method is equivalent to call 
        :obj:`~shapelets.compute.random.ShapeletsRandomEngine.uniform` on the engine.

    Returns
    -------
    ShapeletsArray
        An new array instance whose values are drawn from the standard normal distribution.

    Examples
    --------
    A row vector with 5 elements drawn from the uniform distribution:

    >>> import shapelets.compute as sc
    >>> sc.random.random((1,5))
    [1 5 1 1]
        0.1583     0.3712     0.3543     0.6450     0.9675 

    """
    return __random(shape, dtype, engine)


def randint(low: int, high: Optional[int] = None, shape: ShapeLike = (1, 1), dtype: DataTypeLike = 'int32', engine: Optional[ShapeletsRandomEngine] = None) -> ShapeletsArray:
    """
    Return random integers drawn from the “discrete uniform” in the range low (inclusive) to high (exclusive).

    Parameters
    ----------
    low: int
        When `high` parameter is `None`, it indicates the maximum (exclusive) integer for 
        the range `[0, low(`.  

        If `high` is specified, this parameter sets the miminum (inclusive) of the range.  

    high: Optional int.  Defaults to None.
        When set, determines the highest (exclusive) value for the range.

    shape: ShapeLike.  Defaults to (1,1).
        Shape of the array to be returned.

    dtype: DataTypeLike.  Defaults to int32.
        Final type of elements of the array.

    engine: ShapeletsRandomEngine.  Defaults to None.
        When not set, uses the default, implicit, random engine algorithm.  However,
        when it is set, this method is equivalent to call 
        :obj:`~shapelets.compute.random.ShapeletsRandomEngine.uniform` on the engine adjusting
        the parameters and type for the correct result.

    Returns
    -------
    ShapeletsArray
        An new array instance whose values are drawn from the uniform distribution.

    Examples
    --------
    3 by 3 matrix whose elements range from 0 to 10 (exclusive).

    >>> import shapelets.compute as sc
    >>> sc.random.randint(10, shape=(3,3))
    [3 3 1 1]
            6          2          2 
            0          0          3 
            9          5          7    

    Random row vector of 5 elements in the range ``[10,20(``

    >>> import shapelets.compute as sc
    >>> sc.random.randint(10, 20, (1,5))
    [1 5 1 1]
            16         10         19         12         10 

    """
    return __randint(low, high, shape, dtype, engine)


def permutation(x: ArrayLike, axis: int = 0, engine: Optional[ShapeletsRandomEngine] = None) -> ShapeletsArray:
    r"""
    Randomly shuffles an array around a particular axis.

    Parameters
    ----------
    x: ArrayLike.  Defaults to (1,1).
        Input array

    axis: int.  Defaults to 0
        Axis to be shuffled

    engine: ShapeletsRandomEngine.  Defaults to None.
        When not set, uses the default, implicit, random engine algorithm.  However,
        when it is set, this method is equivalent to call 
        :obj:`~shapelets.compute.random.ShapeletsRandomEngine.uniform` on the engine adjusting
        the parameters and type for the correct result.

    Returns
    -------
    ShapeletsArray
        A randomly suffled array.

    Examples
    --------
    Given the following 2 by 5 matrix

    >>> import shapelets.compute as sc
    >>> a = sc.array([[1,2,3,4,5],[6,7,8,9,0]])
    [2 5 1 1]
            1          2          3          4          5 
            6          7          8          9          0 

    One could reshuffle its columns by

    >>> sc.random.permute(x, 1)        
    [2 5 1 1]
            2          5          4          1          3 
            7          0          9          6          8 

    """
    return __permutation(x, axis, engine)


__all__ = [
    "random_engine",
    "randint",
    "randn",
    "random",
    "permutation",
    "ShapeletsRandomEngine",
    "RandomEngineType"
]
